RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT_vect.cc

```diff
@@ -0,0 +1,1030 @@
+  constexpr double epsilont=1e-2;      // minimum split size t
+  unsigned int nv = y.getSize();
+  
+  // avoid left-right biases by splitting highest Tc first
+  
+  std::vector<std::pair<double, unsigned int> > critical;
+  for(unsigned int k=0; k<nv; k++){    
+    double Tc= 2*y.swE_[k]/y.sw_[k];
+    if (beta*Tc > threshold){
+      critical.push_back( make_pair(Tc, k));
+    }
+  }
+  if (critical.empty()) return false;
+
+
+  std::stable_sort(critical.begin(), critical.end(), std::greater<std::pair<double, unsigned int> >() );
+  
+  
+  bool split=false;
+  const unsigned int nt = tks.getSize();
+
+  for(unsigned int ic=0; ic<critical.size(); ic++){
+    unsigned int k=critical[ic].second;
+
+    // estimate subcluster positions and weight
+    double p1=0, z1=0, t1=0, w1=0;
+    double p2=0, z2=0, t2=0, w2=0;
+    for(unsigned int i=0; i<nt; ++i){
+      if (tks.Z_sum_[i] > 1.e-100) {
+	// winner-takes-all, usually overestimates splitting
+	double tl = tks.z_[i] < y.z_[k] ? 1.: 0.;
+	double tr = 1. - tl;
+
+	// soften it, especially at low T
+	/*
+	double arg = ( tks.z_[i] - y.z_[k] ) * sqrt(beta * tks.dz2_[i]); // + std::fabs(tks.t_[i] - y.t_[k])
+	std::cout << arg << std::endl;
+	if(std::fabs(arg) < 20){
+	  double t = local_exp(-arg);
+	  tl = t/(t+1.);
+	  tr = 1/(t+1.);
+	}
+	*/
+
+	double p = y.pk_[k] * tks.pi_[i] * local_exp(-beta * Eik(tks.z_[i], y.z_[k], tks.dz2_[i], 
+                                                                 tks.t_[i], y.t_[k], tks.dt2_[i])) / tks.Z_sum_[i];
+	double w = p*tks.errsum_[i];
+	p1 += p*tl;  z1 += w*tl*tks.z_[i]; t1 += w*tl*tks.t_[i]; w1 += w*tl;
+	p2 += p*tr;  z2 += w*tr*tks.z_[i]; t2 += w*tr*tks.t_[i]; w2 += w*tr;
+      }
+    }
+
+    if(w1>0){z1 = z1/w1; t1 = t1/w1;} else {z1=y.z_[k]-epsilonz; t1=y.t_[k]-epsilont;}
+    if(w2>0){z2 = z2/w2; t2 = t2/w2;} else {z2=y.z_[k]+epsilonz; t2=y.t_[k]+epsilont;}
+    
+    // reduce split size if there is not enough room
+    if( ( k   > 0 ) && ( z1 < (0.6*y.z_[k] + 0.4*y.z_[k-1])) ){ 
+      z1 = 0.5*y.z_[k] + 0.5*y.z_[k-1]; 
+      t1 = 0.5*y.t_[k] + 0.5*y.t_[k-1]; 
+    }
+    if( ( k+1 < nv) && ( z2 > (0.6*y.z_[k] + 0.4*y.z_[k+1])) ){ 
+      z2 = 0.5*y.z_[k] + 0.5*y.z_[k+1]; 
+      t2 = 0.5*y.t_[k] + 0.5*y.t_[k+1]; 
+    }
+    
+#ifdef VI_DEBUG
+    if(verbose_){
+      if (std::fabs(y.z_[k] - zdumpcenter_) < zdumpwidth_){
+	std::cout << " T= " << std::setw(8) << 1./beta 
+		  << " Tc= " << critical[ic].first 
+		  << "    splitting " << std::fixed << std::setprecision(4) << y.z_[k] 
+		  << " --> (" << z1 << ',' << t1<< "),(" << z2 << ',' << t2 
+		  << ")     [" << p1 << "," << p2 << "]" ;
+	if (std::fabs(z2-z1) > epsilonz || std::fabs(t2-t1) > epsilont){
+	  std::cout << std::endl;
+	}else{
+	  std::cout <<  "  rejected " << std::endl;
+	}
+      }
+    }
+#endif
+
+    // split if the new subclusters are significantly separated
+    if( std::fabs(z2-z1) > epsilonz || std::fabs(t2-t1) > epsilont){
```

Comment:
There was no abs in the Z_vect version.
Was it OK there (due to ordering in 1D), or was it actually a bug?

Reply:
That was a bug and I left it. The vertices are ordered by critical temperature anyway.

Reply:
OK, I made a comment there in https://github.com/cms-sw/cmssw/pull/19935#ref-pullrequest-221269469